### ISSUE TITLE
Required file not found

### ISSUE THREAD

**Issue by Secnd** [Zk4-ps/EaaS-Embedding-Watermark#4]
The project appears to be missing the "attack_autolength.py" file, I cannot locate it in the repository.

**Comment by Zk4-ps** [comment]
I apologize for my negligence, I have uploaded the "attack_autolength.py" file to the directory "semantic_aware_watermark/preparation". This file contains several function used for the disturb embeddings sampling after obtaining the optimal suffixes. We use the third party API platform for getting the embeddings (It's the same as the OpenAI API). If you really need the disturb embedding files, just let me know❤️. I hope this can help you.

**Comment by Secnd** [comment]
Thanks for your reply.I do need these files to conduct some experiments, and it would be wonderful if you could provide them to me.

**Comment by Zk4-ps** [comment]
I have already upload the perturbed text embeddings and the result files of several metrics. You can find them via the link in readme. Hope this can help you.

**Comment by Secnd** [comment]
Thanks a lot for your help! Still, I’m running into some issues with SAW. In extract_emb.py, I can’t import DATA_INFO. I found a similar definition in run_gpt_backdoor.py under embamaer_attack, but the dict in the function LoadtoJson  of extract_emb.py requires a key 'label' , which isn’t mentioned in the earlier dict. I’m a bit lost on the details—could you please point me in the right direction to fix this?

**Comment by Zk4-ps** [comment]
I apologize for my negligence😭, I have add the DATA_INFO in attack_autolength.py. The DATA_INFO value is consistent with EmbMarker. Thus, if you want to extract the embedding yourself, you will have to also use the emb_cache from EmbMarker.

**Comment by Secnd** [comment]
While training the encoder and decoder in the SAW module, I noticed that the training file was set as: `train_file = os.path.join(args.data_dir, 'pf_wm_train_emb.json'`.  So, I experimented by using the embeddings from the SST2 dataset in the EmbMarker within extract_emb.py to generate a json file for training. However, I discovered that the generated file is a list of dict, while the constructor of CustomDataset in custom_dataset.py expects a single dict. I'm a bit confused,did I do something wrong?

**Comment by FZaKK** [comment]
Well, I think you maybe use the file with the format [{'idx', 'label', `'text'}]. The correct file is like "data/sst2/train_emb.json" with the format {'idx': 'embedding'}. The main target for SAW is to inject the watermark into the embeddings. Thus, the embbedings are the inputs. Just be free of `'pf_wm_train_emb.json'`, it's is a temp file name.

**Comment by Zk4-ps** [comment]
Well, I think you maybe use the file with the format `[{'idx', 'label', 'text'}]`. The correct file is like `"data/sst2/train_emb.json"` with the format `{'idx': 'embedding'}`. The main target for SAW is to inject the watermark into the embeddings. Thus, the embbedings are the inputs. Just be free of `'pf_wm_train_emb.json'`, it's is a temp file name.

**Comment by Secnd** [comment]
Got it! Now I wanna test the performence of SAW based on the embeddings you provided(from `Data4SPA`).I have already trained the model. When I execute `encoder_watermark.py`, I don’t know what the parameters `test_disturb_path `and `subset_result_path` are. Did I miss something? Could you how should I obtain them?

**Comment by Zk4-ps** [comment]
`encoder_watermark.py` is prepared for testing the watermark injection model's performance also using the cos, L2 and PCA AUPRC metrics. Thus, it's just like the `attack_warden.py`. And `test_disturb_path = f'data/{dataset}/standard_train_subset_result_emb.json'` \&  `subset_result_path = f'data/{dataset}/standard_train_subset_result.json'`. The `test_disturb_path` is the disturb embeddings files from `Data4SPA` and the `subset_result_path` is the semantic disturb result files from `Data4SPA`.

**Comment by Secnd** [comment]
Thank you very much for answering my doubts so many times, which has helped me better understand the thought process of the paper and the details of its code implementation. I have one last question: when I verify the SAW, my process is to first run `extract_emb.py` to obtain the training data, then run `main.py` to get the model, and finally use `watermark_verification.py `to verify it. Taking SST-2 as an example, I replaced the message according to the log records, but it produced very strange results:delta cos: [[-7.3313713e-06]]
delta L2: [-0.00012732]
delta bit error: 0.0
ks p-value: 1.0

**Comment by Zk4-ps** [comment]
Well, it's a little bit strange. I think your process is correct. I will show you my own experiment results on SST2. 

```
delta cos: [[1.3140045]]
delta L2: [-10.294491]
delta bit error: -12.010745755426607
ks p-value: 5e-324
```

Because of the simplicity of the model, thus the model is suitable for the length of watermark vector under 48 bit. And I will provide a training log example in the repo (/semantic_aware_watermark/log), maybe some hyperparameters or training processes are different. Hope this example file will help you. If it still doesn't work, just let me know.

**Comment by Secnd** [comment]
I checked my log file, and the training crashed at the 19th epoch. I'm not sure if this was a random occurrence or if something went wrong on my end. Do I need to lower the number of epochs when training?Epoch 19 training duration 1.05 sec
----------------------------------------
Running validation for epoch 19/100
loss                   204302027980.8000
encoder_mse            204302027980.8000
encoder_cos            -0.0022
encoder_l2             17702872.4000
decoder_bce            41.6667
decoder_cos            0.5455
decoder_l2             3.1623
bitwise-error          0.3125

**Comment by Secnd** [comment]
I checked my log file, and the training crashed at the 19th epoch. I'm not sure if this was a random occurrence or if something went wrong on my end. Should I  lower the epochs when training?Here is the result:
Running validation for epoch 19/100
loss                   204302027980.8000
encoder_mse            204302027980.8000
encoder_cos            -0.0022
encoder_l2             17702872.4000
decoder_bce            41.6667
decoder_cos            0.5455
decoder_l2             3.1623
bitwise-error          0.3125

**Comment by Secnd** [comment]
I checked my log file, and the training crashed around the 20th epoch, and this happened with other datasets as well. I’m not sure if this is random or if something went wrong on my end. Do I need to lower the number of epochs when training?Here is an log of sst2:Running validation for epoch 19/100
loss                   204302027980.8000
encoder_mse            204302027980.8000
encoder_cos            -0.0022
encoder_l2             17702872.4000
decoder_bce            41.6667
decoder_cos            0.5455
decoder_l2             3.1623
bitwise-error          0.3125

**Comment by Zk4-ps** [comment]
If the training process only includes 20 epochs, the model may not converge, and 100 epochs is an appropriate number. Maybe you can check the default value of args.epochs, or set the epochs with the bash commands `python main.py new --name (dataset name) --data-dir (dataset directory) --batch-size 32 --epochs ...`. If it still crashed around the 20th epoch, this may not be an error with the program code🤔. I run my own experiments on the  server with the NVIDIA GeForce RTX 4090 (CUDA Version: 12.2).

**Comment by Secnd** [comment]
I will attempt to investigate the cause. Furthermore, I'm a bit confused about the term 'original' in Table 3 of Section 6.3 in your paper. Could you clarify its meaning?

**Comment by Zk4-ps** [comment]
Well, we only inject the watermark informantion into the text embeddings with triggers. Thus, the verify text dataset could be divided into two parts according to triggers (with or without triggers). Thus, the term 'original' in Table 3 means the difference between two parts if we don't inject the watermark into the embeddings. This indicates that the original case cannot be verified without injecting the watermark.
